File: scala/src/main/scala/hearth/crypto/Ed25519.scala
package hearth.crypto

/** Ed25519 (EdDSA, RFC 8032) keys and signatures — the standard, Ledger-native
  * signature scheme. The same key material also backs the ECVRF (see [[Ecvrf]]).
  *
  * All operations run on a [[CryptoBackend]] (libsodium or pure-JVM), taken as a
  * `using` parameter that defaults to [[Crypto]].
  */
object Ed25519:

  /** An Ed25519 keypair derived from a 32-byte seed.
    *
    *  - `seed`       : the 32-byte SLIP-0010 node key (the RFC 9381 "SK").
    *  - `publicKey`  : 32-byte compressed point Y = x*B.
    *  - `secretKey`  : the 64-byte expanded form (seed || publicKey).
    */
  final case class KeyPair(seed: Array[Byte], publicKey: Array[Byte], secretKey: Array[Byte]):
    def sign(message: Array[Byte])(using b: CryptoBackend): Array[Byte] = b.signDetached(message, secretKey)

  def fromSeed(seed: Array[Byte])(using b: CryptoBackend): KeyPair =
    require(seed.length == 32, "Ed25519 seed must be 32 bytes")
    val (pk, sk) = b.signSeedKeypair(seed)
    KeyPair(seed, pk, sk)

  def verify(signature: Array[Byte], message: Array[Byte], publicKey: Array[Byte])(using b: CryptoBackend): Boolean =
    b.verifyDetached(signature, message, publicKey)

  /** RFC 8032 secret scalar: clamp(SHA-512(seed)[0..32]). Used by ECVRF. */
  private[crypto] def secretScalar(seed: Array[Byte])(using b: CryptoBackend): Array[Byte] =
    val h = b.sha512(seed)
    val a = h.slice(0, 32)
    a(0)  = (a(0) & 0xf8).toByte
    a(31) = ((a(31) & 0x7f) | 0x40).toByte
    a

  /** RFC 8032 nonce prefix: SHA-512(seed)[32..64]. Used by ECVRF nonce gen. */
  private[crypto] def noncePrefix(seed: Array[Byte])(using b: CryptoBackend): Array[Byte] =
    b.sha512(seed).slice(32, 64)

File: java/src/main/java/tech/hearth/crypto/Ed25519.java
package tech.hearth.crypto;

import java.util.Arrays;

/**
 * Ed25519 (EdDSA, RFC 8032) signature verification and shared helpers. The
 * secret-holding key types live in {@link SigningKey} and {@link VrfKey}; this
 * class only verifies signatures (by public key) and exposes the RFC 8032 scalar
 * derivation that the ECVRF shares.
 */
public final class Ed25519 {
    private Ed25519() {}

    /** Verify a detached signature using the default backend. */
    public static boolean verify(byte[] signature, byte[] message, byte[] publicKey) {
        return verify(signature, message, publicKey, Crypto.defaultBackend());
    }

    public static boolean verify(byte[] signature, byte[] message, byte[] publicKey, CryptoBackend backend) {
        return backend.verifyDetached(signature, message, publicKey);
    }

    /** RFC 8032 secret scalar: clamp(SHA-512(seed)[0..32]). Used by ECVRF. */
    static byte[] secretScalar(byte[] seed, CryptoBackend backend) {
        byte[] a = Arrays.copyOfRange(backend.sha512(seed), 0, 32);
        a[0] &= 0xF8;
        a[31] = (byte) ((a[31] & 0x7F) | 0x40);
        return a;
    }

    /** RFC 8032 nonce prefix: SHA-512(seed)[32..64]. Used by ECVRF nonce gen. */
    static byte[] noncePrefix(byte[] seed, CryptoBackend backend) {
        return Arrays.copyOfRange(backend.sha512(seed), 32, 64);
    }

    /**
     * The X25519 public key sharing this Ed25519 public key's point, under the
     * standard birational map (libsodium's
     * {@code crypto_sign_ed25519_pk_to_curve25519}). Used by {@link SigningKey#toX25519()}.
     */
    static byte[] toX25519PublicKey(byte[] ed25519PublicKey) {
        Ed25519Math.Point point = Ed25519Math.decode(ed25519PublicKey);
        if (point == null) {
            throw new IllegalArgumentException("not a valid Ed25519 public key");
        }
        return Ed25519Math.montgomeryU(point.y());
    }
}
